### PR TITLE
feat(invested): add invested history

### PR DESCRIPTION
**Pull request by codeofsumit**
Add the history of invested capital per day.
Also adds a new utility function to enrich sell-activities with the original purchaseAmount. Very useful for different purposes
This will enable a value chart that includes invested amounts like this:

image

### CODE REVIEW

File: src/calcValueHistory.js [hunk 1]
```diff
@@ -31,18 +31,10 @@ module.exports = function (activities, quotes, interval, i) {
...
-    .map((a) => {
-      return {
-        ...a,
-        amount: a.type === 'Sell' ? a.amount * -1 : a.amount,
-      };
-    });
-
   // calc invested capital at the start of the interval
-  let investedStorage = sumBy(activitiesForInvestedValueBeforeInterval, 'amount');
+  const { purchases: purchasesBeforeInterval } = calcInventoryPurchasesFIFO(activitiesBeforeInterval);
```
Comment (kevcodez) [inline src/calcValueHistory.js]: You know what it is

[](https://media.giphy.com/media/gw3IWyGkC0rsazTi/giphy.gif)

File: src/calcValueHistory.js [hunk 2]
```diff
@@ -31,18 +31,10 @@ module.exports = function (activities, quotes, interval, i) {
...
+  const investedBeforeInterval = sumBy(purchasesBeforeInterval, 'amount');
+  let investedStorage = investedBeforeInterval;
```
Comment (kevcodez) [inline src/calcValueHistory.js]: Is this value modified further down or why do we need to re-assign the constant to a mutable variable (let)?

Comment (codeofsumit) [reply src/calcValueHistory.js]: you were too fast. This was an old try and is now removed

**Comment by codeofsumit** [comment]
Regarding tests: this repo has no test setup yet. I would test the function in tresor-api until we have tests in here.
Personally, I'm not a fan of this repo anymore. Maybe we should just move everything into the API utils 🤔

File: src/calcInvestedCapitalHistory.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+module.exports = function (activities, interval) {
+  if (activities.length === 0) {
+    return {
+      capitalHistory: [],
+      dates: [],
+    };
+  }
+
+  // adjust shares bought/sold by splits that happened in the past
+  activities = applySplitMultiplier(activities);
```
Comment (kevcodez) [inline src/calcInvestedCapitalHistory.js:16]: Please avoid re-assignments of parameters and rather use new variables

File: utils.js [hunk 4]
```diff
@@ -94,7 +95,70 @@ function getPreviousValue(arr, i) {
   }
 }
 
+function calcSalesDataFIFO(activities) {
+  // order asc - it's necessary for a FIFO loop. earliest/oldest activity is first in the array. Latest/newest is last
+  activities = orderBy(cloneDeep(activities), 'date', 'asc');
```
Comment (kevcodez) [inline utils.js:100]: Let's avoid parameter re-assignments and rather use new variables

File: utils.js [hunk 5]
```diff
@@ -94,7 +95,70 @@ function getPreviousValue(arr, i) {
   }
 }
 
+function calcSalesDataFIFO(activities) {
```
Comment (kevcodez) [inline utils.js:98]: We don't have to do it in this PR, but we should definitely add JSDoc to make this easier to use (no need to have a mental model of every return type and parameter). But we can also do this when moving the code to the API repository.

**Comment by kevcodez** [comment]
If this repository isn't used for transparency/contribution and it's not used in multiple projects (i.e. API and frontend), having all of this code separately and untested adds additional complexity, compared to moving it to the API repo, so 👍  for getting rid of this separate repo.

**Comment by kevcodez** [comment]
Good amount of comments, I like that. We should add JSDoc too, but it doesn't have to be introduced in this PR (maybe after moving repo).

**Review by kevcodez — APPROVED** [review]

File: utils.js (hunk 5 above)
Comment (codeofsumit) [reply utils.js:98]: Well do that as part of moving this repo to the API repo 👍

File: src/calcInvestedCapitalHistory.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+module.exports = function (activities, interval) {
+  if (activities.length === 0) {
+    return {
+      capitalHistory: [],
+      dates: [],
+    };
+  }
+
+  // adjust shares bought/sold by splits that happened in the past
+  const tempActivities = applySplitMultiplier(activities);
```
Comment (kevcodez) [inline src/calcInvestedCapitalHistory.js:16]: What about renaming the var to something that matches the content? Like "activitiesWithSplitMultiplier". Words like "temp" or "adjusted" don't really add any value to the variable and I couldn't tell what it is without looking up the calling function.

**Review by kevcodez — APPROVED** [review]

**Pull request merged**